New block: Post comments link

carolinan opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->

Part of https://github.com/WordPress/gutenberg/issues/22724
Adds a new post comments link block.

The block displays the number of comments, and links to the comments area on the single view.

❓  **-It tries to link to #comments even if the single view does not have a comment(s) block.**

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

I have added the block inside the post content, and viewed the post on the front:
With and without comments, with and without comments enabled.

I have added the block to the index.html file of TT1 Blocks, to make sure that it inherits the context and displays correctly in a query loop.

## Screenshots <!-- if applicable -->
TT1 Blocks, front:
![image](https://user-images.githubusercontent.com/7422055/109969141-d6593600-7cf3-11eb-8cd5-571e14daa932.png)
_The title is showing because the theme does not have a screen-reader-text class._

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
New block.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

carolinan committed:
New block: Post comments link

carolinan commented:
I haven't learnt how to write the tests yet, saving that for tomorrow 😸

carolinan committed:
Linting

carolinan committed:
Linting

carolinan committed:
one more try

carolinan committed:
Add fixtures

Add the generated fixtures for post-comments-link.

carolinan committed:
Merge pull request #7 from WordPress/trunk

sync

carolinan committed:
Merge branch 'trunk' into add/post-comments-link

carolinan committed:
Fix linting after solving merge conflict

carolinan committed:
Merge pull request #12 from WordPress/trunk

Sync with trunk

scruffian commented:
This looking good. I think it would be good to have an option to customize the text - I might like it to say "Leave a comment" rather than show the number.

It might also be nice to add an option to display an icon next to the link.

jasmussen commented:
This is what I see:

<img width="789" alt="Screenshot 2021-04-05 at 09 19 18" src="https://user-images.githubusercontent.com/1204802/113548603-1ca1fd80-95f0-11eb-938a-33ff549320e6.png">

We need a new icon. But I think that's probably something that can wait, and not something that should block this PR. I'd also echo Ben's comment above about a custom phrasing, but also suggest that's something that can wait (I'd like the same customizability for the pagination blocks, for example).

One thing I noticed is that the block isn't inert:

![focusable](https://user-images.githubusercontent.com/1204802/113548699-49561500-95f0-11eb-85ac-2899fd980f39.gif)

I could swear there was precedence in some existing block for making links like that inert using a `Disabled` component, but I can't find it now, and I'm noticing also Latest Posts and others are interactive. So probably that should also not be a blocker.

So I guess this just needs a code review. But would be good to ticket followups. Thank you!

carolinan commented:
If the disabled component is used we won't be able to also have an editable text.

carolinan committed:
Merge pull request #13 from WordPress/trunk

Sync

jasmussen commented:
> If the disabled component is used we won't be able to also have an editable text.

Good point, though we could just put a textfield in the sidebar. While normally putting it directly in the canvas is the best strategy, I have a vague instinct that the sidebar might be where people intuit a rename to be. But this is not a strong opinion, and as noted, it could definitely be a followup.

carolinan commented:
The use of links inside the editors may need a more comprehensive review so that the behavior is the same in different blocks.
And if there already are best practices, they need to be documented.
The problem with having the link clickable here is that in the site editor, the linked page is opened inside the iframe.
It can also be confusing with visible links that are not clickable (but can be opened with right click), but it seems like the least bad option.

jasmussen commented:
Agreed. We've been talking about a "navigation cursor" you can select from the tools drop-down, which would navigate to link destinations inside the editor, when links are clicked. Sometimes referred to as "browse mode".

aristath commented on packages/block-library/src/post-comments-link/edit.js:
```diff
@@ -0,0 +1,112 @@
+
+	let commentsText;
+	if ( commentsCount !== undefined ) {
+		if ( commentsCount.length === 0 ) {
+			commentsText = __( 'No comments' );
+		} else if ( commentsCount.length === 1 ) {
+			commentsText = __( 'One comment' );
+		} else {
+			commentsText = sprintf(
+				/* translators: %d Number of comments */
+				__( '%d comments' ),
+				commentsCount
+			);
+		}
```
Shouldn't this be using `_n()`? That would account for languages where the plural is not the same for all numbers (in some languages the plural for 2 is different)

carolinan replied:
Did you mean like this?
```
if ( commentsCount.length === 0 ) {
		commentsText = __( 'No comments' );
	} else {
		commentsText = sprintf(
		/* translators: %d Number of comments */
		_n(
			'%d comment',
			'%d comments',
			commentsCount
		),
		commentsCount
	);
}

```

carolinan replied:
Ok that did not work because it now says "1 comments"

carolinan replied:
The `commentsCount.length i`s also incorrect here

aristath commented on packages/block-library/src/post-comments-link/index.php:
```diff
@@ -0,0 +1,67 @@
+		|| isset( $block->context['postId'] )
+		&& ! comments_open( $block->context['postId'] ) ) {
+			return '';
+	}
```
Nitpicking, but could we change the alignment of this to make things line-up a bit more, and have `return` nested 1 level less?

```suggestion
	if (
		! isset( $block->context['postId'] ) ||
		isset( $block->context['postId'] ) &&
		! comments_open( $block->context['postId'] )
	) {
		return '';
	}
```

aristath commented on packages/block-library/src/post-comments-link/index.php:
```diff
@@ -0,0 +1,67 @@
+	$post_title         = get_the_title( $block->context['postId'] );
+	$comment_text       = '';
+
+	if ( 0 === $comments_number ) {
+		$comment_text = sprintf(
+			/* translators: %s post title */
+			__( 'No comments<span class="screen-reader-text"> on %s</span>' ),
+			$post_title
+		);
+	} elseif ( 1 === $comments_number ) {
+		$comment_text = sprintf(
+			/* translators: %s post title */
+			__( 'One comment<span class="screen-reader-text"> on %s</span>' ),
+			$post_title
+		);
+	} else {
+		$comment_text = sprintf(
+			/* translators: %1$d Number of comments, %2$s post title */
+			__( '%1$d comments<span class="screen-reader-text"> on %2$s</span>' ),
+			$comments_number,
+			$post_title
+		);
+	}
```
This should probably be using `_n()` too.

aristath commented on packages/block-library/src/post-comments-link/index.php:
```diff
@@ -0,0 +1,67 @@
+		);
+	}
+
+	return sprintf( '<div %1$s>', $wrapper_attributes ) .
+		'<a href="' . $comments_link . '">' . $comment_text . '</a>' .
+		'</div>';
```
A bit of a nitpick, but `sprintf()` is not really necessary here, and we should avoid concatenating strings like `'</a>' . '</div>'` when they can be combined.

```suggestion
	return "<div {$wrapper_attributes}><a href='{$comments_link}'>{$comment_text}</a></div>";
```

carolinan committed:
Set text color to false

Set text color to false in block.json,
Updated comment count logic, update index.php

carolinan committed:
Merge pull request #14 from WordPress/trunk

Sync

carolinan committed:
Use _n()

aristath commented on packages/block-library/src/post-comments-link/edit.js:
```diff
@@ -0,0 +1,116 @@
+		} else {
+			commentsText = sprintf(
+				/* translators: %s: Number of comments */
+				_n(
+					'%s comment',
+					'%s comments',
+					commentsNumber
+				),
```
The linter is complaining, this needs to change to the below:
```suggestion
				_n( '%s comment', '%s comments', commentsNumber ),
```

carolinan committed:
fix linting
